qaoa.py

```diff
@@ -130,3 +146,204 @@ def expectationValue_MaxCut(data,G):
             y=[int(i) for i in binstring]
             E[item] += cost_MaxCut(y,G)*c/shots
     return E
+
+
+
+def objective_function(params, G):
+    """
+    :return: minus the expectation value (in order to maximize MaxCut configuration)
+    NB! If a list of circuits are ran, only returns the expectation value of the first circuit.
+    """
+    qc = createCircuit_MaxCut(params, G, int(len(params)/2))
+    res_data = execute(qc, backend, shots=NUM_SHOTS).result().results
+    E = expectationValue_MaxCut(res_data, G)
+    return -E[0]
+
+def random_init(depth, weighted):
+    """
+    Enforces the bounds of gamma and beta based on the graph type.
+    :return: np.array on the form (gamma_1, beta_1, gamma_2, ...., gamma_d, beta_d)
+    """
+    if weighted:
+        gamma_list = np.random.uniform(-10,10, size=depth) # Here 10 is arbitrary
+    else:
+        gamma_list = np.random.uniform(-np.pi / 2, np.pi / 2, size=depth)
+    beta_list = np.random.uniform(-np.pi / 4, np.pi / 4, size=depth)
+    initial = np.empty((gamma_list.size + beta_list.size,), dtype=gamma_list.dtype)
+    initial[0::2] = gamma_list
+    initial[1::2] = beta_list
+    return initial
+
+def get_constaints_for_COBYLA(depth, weighted):
+    """
+    :return: List of constraints applying to the parameters
+    """
+    bounds = []
+    if weighted:
+        for i in range(depth):
+            bounds.append([-10,10]) # Arbitrary choice
+            bounds.append([-np.pi/4, np.pi/4])
+    else:
+        for i in range(depth):
+            bounds.append([-np.pi/2, np.pi/2])
+            bounds.append([-np.pi/4, np.pi/4])
+    cons = []
+    for factor in range(len(bounds)):
+        lower, upper = bounds[factor]
+        l = {'type': 'ineq',
+             'fun': lambda x, lb=lower, i=factor: x[i] - lb}
+        u = {'type': 'ineq',
+             'fun': lambda x, ub=upper, i=factor: ub - x[i]}
+        cons.append(l)
+        cons.append(u)
+    return cons
+
+def optimize_random(K,G, depth=1, weighted=False):
+    """
+    :param K: # Random initializations (RIs)
+    :return: Array of best params (on the format where the gammas and betas are intertwined),
+    the corresponding best energy value, and the average energy value for all the RIs
+    """
+    rec = -np.inf
+    avg_list = np.zeros(K)
+    rec_params = 42 # For debug purposes
+    for i in range(K):
+        init_params = random_init(depth, weighted)
+        cons = get_constaints_for_COBYLA(depth, weighted)
+        sol = minimize(objective_function, x0=init_params, method='COBYLA', args=(G), constraints=cons)
+        params = sol.x
+        qc = createCircuit_MaxCut(params, G, depth)
+        temp_res_data = execute(qc, backend, shots=NUM_SHOTS).result().results
+        E = expectationValue_MaxCut(temp_res_data, G)[0]
+        avg_list[i] = E
+        if E>rec:
+            rec = E
+            rec_params = params
+    return rec_params, rec, np.average(avg_list)
```

Comment:
what does "rec" stand for?

Reply:
record. I will make the change from rec -> record